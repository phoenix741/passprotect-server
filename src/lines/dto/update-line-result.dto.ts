import { createUnionType } from 'type-graphql';
import { Errors } from '../../shared/dto/errors.dto';
import { WalletLine } from './wallet-line.dto';

export const UpdateLineResultUnion = createUnionType({
  name: 'UpdateLineResult',
  description: 'Result of creation or update of the line',
  types: [WalletLine, Errors],
  resolveType: (obj: WalletLine | Errors): string => {
    const hasError = (tbd: WalletLine | Errors): tbd is Errors => !!(tbd as Errors).errors;
    if (hasError(obj)) {
      return 'Errors';
    }
    return 'WalletLine';
  },
});

export type UpdateLineResult = typeof UpdateLineResultUnion;
