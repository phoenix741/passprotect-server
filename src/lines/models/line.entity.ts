import { Document } from 'mongoose';
import { EncryptedContent } from '../../shared/interfaces/encrypted-content.interface';
import { LineTypeEnum } from '../../shared/interfaces/line-type-enum.interface';
import { ObjectID } from 'bson';

export interface LineEntity extends Document {
  _id: ObjectID;
  type: LineTypeEnum;
  label: string;
  group: string;
  logo?: Buffer;
  encryption: EncryptedContent;
  user: string;
  createdAt: Date;
  updatedAt: Date;
  _rev: number;
}
