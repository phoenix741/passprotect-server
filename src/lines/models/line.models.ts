import { EncryptedContent } from '../../shared/interfaces/encrypted-content.interface';
import { LineTypeEnum } from '../../shared/interfaces/line-type-enum.interface';
import { ObjectID } from 'bson';

export interface LineToCreate {
  label: string;
  type: LineTypeEnum;
  group: string;
  encryption: EncryptedContent;
  user: string;
}

export interface LineToUpdate {
  _id?: ObjectID;
  label: string;
  type: LineTypeEnum;
  group: string;
  encryption: EncryptedContent;
}
