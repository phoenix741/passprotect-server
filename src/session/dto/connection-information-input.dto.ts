import { InputType, Field } from 'type-graphql';

@InputType({ description: 'Information needed to connect to the server and retreive object from database.' })
export class ConnectionInformationInput {
  @Field({ description: 'Login of the user' })
  username: string;

  @Field({ description: 'Password used for connection' })
  password: string;
}
