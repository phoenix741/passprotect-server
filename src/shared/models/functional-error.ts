import { HttpException } from '@nestjs/common';

export class FunctionalError extends Error {
  field?: string;

  constructor(field?: string, message?: string) {
    super(message);

    this.field = field;
  }

  toGraphQL(defaultLabel = 'label') {
    return {
      errors: [{
        fieldName: this.field || defaultLabel,
        message: this.message,
      }],
    };
  }
}

export function toFunctionalError(error: Error): FunctionalError {
  if (error instanceof FunctionalError) {
    return error;
  }
  if (error instanceof HttpException) {
    return new FunctionalError(undefined, error.message.message);
  }
  return new FunctionalError(undefined, error.message);
}
