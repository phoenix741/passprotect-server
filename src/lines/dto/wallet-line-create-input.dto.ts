import { InputType, Field, ID } from 'type-graphql';
import { EncryptedBlockInput } from '../../shared/dto/encrypted-block-input.dto';
import { LineTypeEnum } from '../../shared/interfaces/line-type-enum.interface';

@InputType({ description: 'Wallet line used for creation' })
export class WalletLineCreateInput {
  @Field(type => LineTypeEnum, {
    description: 'Type of the line (text, password, card)',
  })
  type: LineTypeEnum;

  @Field({ description: 'Label of the wallet line' })
  label: string;

  @Field({ description: 'Group of the wallet line' })
  group: string;

  @Field({ description: 'Logo of the line', nullable: true })
  logo?: string;

  @Field(type => EncryptedBlockInput, {
    description: 'Encrypted content of the line (encrypted client side)',
  })
  encryption: EncryptedBlockInput;
}
