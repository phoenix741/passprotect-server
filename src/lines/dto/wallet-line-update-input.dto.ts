import { InputType, Field } from 'type-graphql';
import { ObjectID } from 'bson';
import { LineTypeEnum } from '../../shared/interfaces/line-type-enum.interface';
import { EncryptedBlockInput } from '../../shared/dto/encrypted-block-input.dto';

@InputType({ description: 'Wallet line used for update' })
export class WalletLineUpdateInput {
  @Field(type => ObjectID, {
    description: 'Id of the line, if not filled application create a new line',
    nullable: true,
  })
  _id: ObjectID;

  @Field(type => LineTypeEnum, {
    description: 'Type of the line (text, password, card)',
    nullable: true,
  })
  type?: LineTypeEnum;

  @Field({ description: 'Label of the wallet line', nullable: true })
  label?: string;

  @Field({ description: 'Group of the wallet line', nullable: true })
  group?: string;

  @Field({ description: 'Logo of the line', nullable: true })
  logo?: string;

  @Field(type => EncryptedBlockInput, {
    description: 'Encrypted content of the line (encrypted client side)',
    nullable: true,
  })
  encryption?: EncryptedBlockInput;
}
