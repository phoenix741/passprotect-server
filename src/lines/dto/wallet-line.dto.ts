import { Field, ObjectType, ID, Int } from 'type-graphql';
import { EncryptedBlock } from '../../shared/dto/encrypted-block.dto';
import { LineTypeEnum } from '../../shared/interfaces/line-type-enum.interface';
import { ObjectID } from 'bson';

@ObjectType({ description: 'Wallet' })
export class WalletLine {
  @Field(type => ObjectID, { description: 'Id of the line' })
  _id: ObjectID;

  @Field(type => LineTypeEnum, { description: 'Type of the line (text, password, card)' })
  type: LineTypeEnum;

  @Field({ description: 'Label of the wallet line' })
  label: string;

  @Field({ description: 'Group of the wallet line', nullable: true })
  group?: string;

  @Field(type => Buffer, { description: 'Logo of the line', nullable: true })
  logo?: Buffer;

  @Field(type => EncryptedBlock, { description: 'Encrypted content of the line (encrypted client side)' })
  encryption: EncryptedBlock;

  @Field(type => Date, { description: 'Created date of the line', nullable: true })
  createdAt?: Date;

  @Field(type => Date, { description: 'Last updated date of the line', nullable: true })
  updatedAt?: Date;

  @Field(type => Int, { description: 'Revision (modification count) of the line' })
  _rev?: number;
}
