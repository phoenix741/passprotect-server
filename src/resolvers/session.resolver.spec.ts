import { Test } from '@nestjs/testing';
import { SessionResolver } from './session.resolver';
import { SessionsService } from '../session/sessions.service';
import { User } from '../users/dto/user.dto';
import { ConnectionInformationInput } from '../session/dto/connection-information-input.dto';
import { FunctionalError } from '../shared/models/functional-error';

describe('SessionResolver', () => {
  let sessionResolver: SessionResolver;
  let sessionService;

  const user: User = new User();

  beforeEach(async () => {
    user._id = 'username';
    sessionService = {
      signIn: jest.fn(),
    };

    const module = await Test.createTestingModule({
      providers: [
        SessionResolver,
        {
          provide: SessionsService,
          useValue: sessionService,
        },
      ],
    }).compile();

    sessionResolver = module.get<SessionResolver>(SessionResolver);
  });

  describe('session', () => {
    it('success', async () => {
      expect(await sessionResolver.session(user)).toMatchSnapshot();
    });
  });

  describe('createSession', () => {
    it('success', async () => {
      const input = new ConnectionInformationInput();
      input.username = 'myUsername';
      input.password = 'password';
      sessionService.signIn.mockImplementation(() => ({
        jwtToken: 'myToken',
        user,
      }));
      expect(await sessionResolver.createSession(input)).toMatchSnapshot();
      expect(sessionService.signIn).toHaveBeenCalledWith('myUsername', 'password');
    });

    it('error - signin', async () => {
      const input = new ConnectionInformationInput();
      input.username = 'myUsername';
      input.password = 'password';
      sessionService.signIn.mockImplementation(() => {
        throw new FunctionalError('myfield', 'mymessage');
      });
      expect(await sessionResolver.createSession(input)).toMatchSnapshot();
      expect(sessionService.signIn).toHaveBeenCalledWith('myUsername', 'password');
    });

    it('error - no username', async () => {
      const input = new ConnectionInformationInput();
      input.username = '';
      input.password = 'password';
      expect(await sessionResolver.createSession(input)).toMatchSnapshot();
      expect(sessionService.signIn).not.toHaveBeenCalled();
    });

    it('error - no password', async () => {
      const input = new ConnectionInformationInput();
      input.username = 'myUsername';
      input.password = '';
      expect(await sessionResolver.createSession(input)).toMatchSnapshot();
      expect(sessionService.signIn).not.toHaveBeenCalled();
    });
  });
});
