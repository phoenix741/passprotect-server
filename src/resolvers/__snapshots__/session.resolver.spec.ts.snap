// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SessionResolver createSession error - no password 1`] = `
Object {
  "errors": Array [
    Object {
      "fieldName": "username",
      "message": "error:user.401.password",
    },
  ],
}
`;

exports[`SessionResolver createSession error - no username 1`] = `
Object {
  "errors": Array [
    Object {
      "fieldName": "username",
      "message": "error:user.401.username",
    },
  ],
}
`;

exports[`SessionResolver createSession error - signin 1`] = `
Object {
  "errors": Array [
    Object {
      "fieldName": "myfield",
      "message": "mymessage",
    },
  ],
}
`;

exports[`SessionResolver createSession success 1`] = `
Object {
  "token": "bearer myToken",
  "user": User {
    "_id": "username",
  },
}
`;

exports[`SessionResolver session success 1`] = `
User {
  "_id": "username",
}
`;
